### PR TITLE
Add an opt-in delay after compute logs finish to ensure that the captured logs are redirected back to stdout/stderr

### PR DESCRIPTION
**Pull request by HynekBlaha**
## Summary & Motivation
https://github.com/dagster-io/dagster/issues/23336

## How I Tested These Changes
- I work on mac (comes with posix tail instead of gnu tail), so I [installed gnu tail](https://apple.stackexchange.com/questions/410513/in-mac-how-to-do-tail-f-pid-to-terminate-after-process-exit-like-linux) to run with `--pid` locally.

```
 ~ > tail --help                                                                                                                                                                  ✔  22:36:02
tail: unrecognized option `--help'
usage: tail [-F | -f | -r] [-q] [-b # | -c # | -n #] [file ...]
 ~ > gtail --help                                                                                                                                                               1 ✘  22:36:06
Usage: gtail [OPTION]... [FILE]...
Print the last 10 lines of each FILE to standard output.
With more than one FILE, precede each with a header giving the file name.

With no FILE, or when FILE is -, read standard input.

Mandatory arguments to long options are mandatory for short options too.
  -c, --bytes=[+]NUM       output the last NUM bytes; or use -c +NUM to
                             output starting with byte NUM of each file
  -f, --follow[={name|descriptor}]
                           output appended data as the file grows;
                             an absent option argument means 'descriptor'
  -F                       same as --follow=name --retry
  -n, --lines=[+]NUM       output the last NUM lines, instead of the last 10;
                             or use -n +NUM to skip NUM-1 lines at the start
      --max-unchanged-stats=N
                           with --follow=name, reopen a FILE which has not
                             changed size after N (default 5) iterations
                             to see if it has been unlinked or renamed
                             (this is the usual case of rotated log files);
                             with inotify, this option is rarely useful
      --pid=PID            with -f, terminate after process ID, PID dies;
                             can be repeated to watch multiple processes
  -q, --quiet, --silent    never output headers giving file names
      --retry              keep trying to open a file if it is inaccessible
  -s, --sleep-interval=N   with -f, sleep for approximately N seconds
                             (default 1.0) between iterations;
                             with inotify and --pid=P, check process P at
                             least once every N seconds
  -v, --verbose            always output headers giving file names
  -z, --zero-terminated    line delimiter is NUL, not newline
      --help        display this help and exit
      --version     output version information and exit
```
- I validated that `--pid` param behaves as expected.
- Discovered there is a `--s` param with default 1.0s. As shown in gif below, the tail process has visible delays. It might be one of reasons why logs were incomplete (tail process was killed from python too fast).
current_interval_1s
- When run with `--s 0.01`, the tail process forwards logs much smoothly. (I would prefer to explicitly set a tighter loop, because I expect the kubernetes container to stop all other processes when main cmd completes).
faster_interval_0 01s

- Build a dev package and installed it to my code deployment.
- Run a dagster job with multiprocessor executor and 5 steps.
- Shelled into dagster job pod and inspected processes.
Screenshot 2024-09-17 at 22 17 45

- ❗️ As tail process finishes with --pid, it keeps hanging as a zombie (`[tail]<defunct>`). I am not 100% sure how big problem it could be, but better raising it here.
- In master branch, it is solved with this piece of code, but I don't want to terminate it (that's why I use --pid). Any advice welcome, @gibsondan.https://github.com/dagster-io/dagster/blob/20e8596e810ded6327052e7c622c39b940241c28/python_modules/dagster/dagster/_core/execution/compute_logs.py#L143

- ✅ `STEP_SUCCESS` events are now visible in grafana (collected from stdout/stderr).

## Changelog

`BUGFIX`(compute_log): incomplete stdout/stderr logs tailed from files

### CODE REVIEW

**Comment by HynekBlaha** [comment]
Would you please find time to review this: @prha, @gibsondan?

**Review by prha — APPROVED** [review]
Is gnu tail doing something smarter around coordinating the output from the logs with the tracked process?  This statement indicate that, but it's not clear what exactly that is:

```
If you specify a pid that is not in use or that does not correspond to the process that is writing to the tailed files, then tail may terminate long before any files stop growing or it may not terminate until long after the real writer has terminated.
```

I'm wondering to see if it's better coordination between the logging process and the tail process in the GNU implementation or the fact that you could adjust the sleep interval.

Either way this PR looks good.

**Comment by gibsondan** [comment]
some thoughts on this after looking into it more:
-I am nervous about moving everybody who is not on mac to gnu tail with no way to opt out - I am seeing some indication that some linux systems do not have it either that our users could plausibly be using, and i'd be worried about breaking them: https://unix.stackexchange.com/questions/760033/alpine-linux-replace-busybox-tail-to-gnu-tail
- I am similarly nervous about making the decreased sleep interval the default, because i think it may increase CPU usage for all of our users' steps (not a ton, but enough that I could see a visible difference when running it locally). I'd be open to making configuring it in an opt in way via an env var?
- I am wondering if fixing this for both POSIX path and GNU path with the default sleep interval could be as simple as adding a line here in the finally block?

```
    finally:
        time.sleep(float(os.getenv("DAGSTER_COMPUTE_LOG_TAIL_WAIT_AFTER_FINISH", "1")))

        if tail_process:
            _clean_up_subprocess(tail_process)

        if watcher_process:
            _clean_up_subprocess(watcher_process)
```

That fixed it for me, and given that the default sleep interval for tail appears to be 1 second in both GNU tail and regular tail, I think 1 second would be a defensible default? It would potentially increase step latency by a second, but ensure that the logs were actually captured, and still be configurable via env var if 1-second latency concerns trump the pod logs being complete.

When I first saw the proposal to use this new command i think i misunderstood the exact benefit - I thought we could use the "--pid" parameter to know for sure when the step process had finished emitting things to stdout, but I now realize that that does not make sense since the pid we are passing in is our own pid.

Let me know what you think about that proposal.

**Review by gibsondan — CHANGES_REQUESTED** [review]
see inline

**Comment by HynekBlaha** [comment]
Hi @gibsondan and @prha, thanks both of you for finding time to review my PR and share your ideas and concerns.

- The gnu/posix tail on different systems than mac is a great catch, that went right over my head. Thank you. 
- My another concern is using the --pid arg and not handling the exit signal of the finished tail process. The count of tail zombie processes would be 2x ops (stdout+stderr). It shouldn't be an issue for kubernetes workloads, but it might be issue for other dagster workload types.
- The sleep interval (0.01) is just way too aggressive.

In conclusion, I don't see my original PR safe. 
As the default sleep interval is 1s, the line with configurable sleep will be much better. 
I reverted the original commit and added another one with the fix @gibsondan suggested.

**Review by gibsondan — COMMENTED** [review]
nice - I kicked off our CI and it looks like one test is failing: https://buildkite.com/dagster/unit-tests/builds/1013#019243b6-64db-414e-a5b5-5094456ff86f - we may just need to increase the timeout in that test, then this is good to go I think.

**Comment by gibsondan** [comment]
Ah sorry i think it is still failing, are you able to reproduce it locally? It's this line that needs a bump I think: https://github.com/dagster-io/dagster/blob/aef8973b2569fa94b24b2c88d1ba2c02c34d9019/python_modules/dagster/dagster_tests/execution_tests/pipes_tests/test_subprocess.py#L757

**Comment by HynekBlaha** [comment]
Yep, I got it wrong the first time, it's fixed now. Thx

File: python_modules/dagster/dagster_tests/execution_tests/pipes_tests/test_subprocess.py [hunk 1]
```diff
@@ -754,7 +754,7 @@ def test_cancellation():
                 p.terminate()
                 break
 
-        p.join(timeout=1)
+        p.join(timeout=3)
```
Comment (HynekBlaha) [inline python_modules/dagster/dagster_tests/execution_tests/pipes_tests/test_subprocess.py:757]: `timeout=2` was not enough when running tests locally:
```python -m pytest python_modules/dagster/dagster_tests -k test_cancellation```

**Comment by HynekBlaha** [comment]
I will run all tests locally...

**Review by gibsondan — COMMENTED** [review]
That test is now passing which is great - the fact that this test suite, which normally finishes within 15 minutes, is now reliably timing out after 25 minutes is giving me pause though: https://buildkite.com/dagster/unit-tests/builds/1017#019244fb-c523-4e4a-adcb-6219d75cfa29

That could be reflective of the fact that the test suite does a bunch of very short ops in-process for testing purposes, neither of which is likely to be an issue that real users face in their jobs - or it could indicate that this PR could cause real perf changes for certain jobs.

One option to move forward quickly would be to keep the env var but default it to "0" instead of "1" to start (i.e. make this fix opt-in initially by setting the env var in your run environment, then hopefully opt-out at a later date once we've done more testing to verify that it won't cause a perf regression)

**Comment by HynekBlaha** [comment]
Sure, opt-in makes total sense. I also added an inline comment to record reasoning behind this. :)

**Review by gibsondan — APPROVED** [review]
ok great - thank you for powering through this! Hopefully we can get enough confidence to make this on in some fashion by default vs. requiring people to rely on an environment variable.
